File: utils/ml_predictor.py
"""
Simple ML predictor for project cost estimation.
Uses historical project data to predict costs based on dimensions.
"""
import os
import json
from typing import Dict, List, Optional
from datetime import datetime

# Try to import ML libraries
try:
    import numpy as np
    from sklearn.linear_model import LinearRegression
    from sklearn.preprocessing import StandardScaler
    ML_AVAILABLE = True
except ImportError:
    ML_AVAILABLE = False
    print("[ML] scikit-learn or numpy not installed. ML features disabled.")

PROJECTS_DIR = ".files"
ML_DATA_FILE = os.path.join(PROJECTS_DIR, "ml_data.json")


class SimplePredictor:
    """Simple cost predictor based on historical data."""
    
    def __init__(self):
        self.training_data = []
        self.model = None
        self.scaler = None
        self.is_trained = False
        self._load_data()
    
    def _load_data(self, auto_train=True):
        """Load training data from file."""
        if os.path.exists(ML_DATA_FILE):
            try:
                with open(ML_DATA_FILE, 'r') as f:
                    self.training_data = json.load(f)
                print(f"[ML] Loaded {len(self.training_data)} training samples")
                if auto_train and len(self.training_data) >= 3 and ML_AVAILABLE:
                    self._train()
            except Exception as e:
                print(f"[ML] Error loading data: {e}")
                self.training_data = []
    
    def _save_data(self):
        """Save training data to file."""
        try:
            os.makedirs(PROJECTS_DIR, exist_ok=True)
            with open(ML_DATA_FILE, 'w') as f:
                json.dump(self.training_data, f, indent=2)
        except Exception as e:
            print(f"[ML] Error saving data: {e}")
    
    def _train(self):
        """Train the model on current data."""
        if not ML_AVAILABLE or len(self.training_data) < 3:
            self.is_trained = False
            return False
        
        try:
            # Extract features: [sqft, bays, width, height]
            X = []
            y = []
            
            for sample in self.training_data:
                features = [
                    sample.get('sqft', 0),
                    sample.get('bays', 0),
                    sample.get('width', 0),
                    sample.get('height', 0)
                ]
                X.append(features)
                y.append(sample.get('cost', 0))
            
            X = np.array(X)
            y = np.array(y)
            
            # Scale features
            self.scaler = StandardScaler()
            X_scaled = self.scaler.fit_transform(X)
            
            # Train simple linear model
            self.model = LinearRegression()
            self.model.fit(X_scaled, y)
            
            self.is_trained = True
            print(f"[ML] Model trained with {len(self.training_data)} samples")
            return True
            
        except Exception as e:
            print(f"[ML] Training error: {e}")
            self.is_trained = False
            return False
    
    def add_sample(self, elevation_data: Dict, project_name: str = "", elevation_name: str = "") -> bool:
        """Add a training sample from elevation data."""
        # Extract cost from material_impact
        material_impact = elevation_data.get('material_impact', [])
        total_cost = sum(m.get('cost_incurred', 0) for m in material_impact)
        
        if total_cost <= 0:
            return False
        
        # Create unique identifier - include elevation name to distinguish same-dimension elevations
        width = elevation_data.get('opening_width_inches', 0)
        height = elevation_data.get('opening_height_inches', 0)
        bays_w = elevation_data.get('bays_wide', 0)
        bays_h = elevation_data.get('bays_tall', 0)
        sqft = elevation_data.get('total_sqft', 0)
        finish = elevation_data.get('finish', '')
        
        # Include elevation name in ID to allow same-dimension elevations
        sample_id = f"{project_name}_{elevation_name}_{width}_{height}_{bays_w}x{bays_h}_{finish}"
        
        # Check for duplicates
        for existing in self.training_data:
            if existing.get('id') == sample_id:
                return False
        
        sample = {
            'id': sample_id,
            'project': project_name,
            'width': width,
            'height': height,
            'bays': bays_w * bays_h,
            'bays_wide': bays_w,
            'bays_tall': bays_h,
            'sqft': sqft,
            'finish': finish,
            'cost': total_cost,
            'timestamp': datetime.now().isoformat()
        }
        
        self.training_data.append(sample)
        self._save_data()
        
        # Retrain if we have enough data
        if len(self.training_data) >= 3:
            self._train()
        
        return True
    
    def predict(self, elevation_data: Dict) -> Dict:
        """Predict cost for an elevation."""
        width = elevation_data.get('opening_width_inches', 0)
        height = elevation_data.get('opening_height_inches', 0)
        bays_w = elevation_data.get('bays_wide', 0)
        bays_h = elevation_data.get('bays_tall', 0)
        sqft = elevation_data.get('total_sqft', 0)
        
        bays = bays_w * bays_h if bays_w and bays_h else 1
        
        # If we have enough training data, check for exact matches first
        for sample in self.training_data:
            if (abs(sample.get('width', 0) - width) < 1 and 
                abs(sample.get('height', 0) - height) < 1 and
                sample.get('bays', 0) == bays):
                return {
                    'predicted_cost': sample['cost'],
                    'confidence': 0.95,
                    'method': 'exact_match',
                    'training_samples': len(self.training_data)
                }
        
        # Use ML model if trained
        if self.is_trained and self.model and ML_AVAILABLE:
            try:
                features = np.array([[sqft, bays, width, height]])
                features_scaled = self.scaler.transform(features)
                predicted = self.model.predict(features_scaled)[0]
                
                # Calculate confidence based on data variance
                costs = [s['cost'] for s in self.training_data]
                avg_cost = sum(costs) / len(costs)
                
                # Higher confidence if prediction is close to average
                diff_pct = abs(predicted - avg_cost) / avg_cost if avg_cost > 0 else 1
                confidence = max(0.4, min(0.85, 1 - diff_pct * 0.5))
                
                return {
                    'predicted_cost': max(0, round(predicted, 2)),
                    'confidence': round(confidence, 2),
                    'method': 'ml_model',
                    'training_samples': len(self.training_data)
                }
            except Exception as e:
                print(f"[ML] Prediction error: {e}")
        
        # Fallback to simple average-based prediction
        if self.training_data:
            # Find most similar by sqft
            avg_cost_per_sqft = sum(s['cost'] / max(1, s['sqft']) for s in self.training_data) / len(self.training_data)
            estimated = sqft * avg_cost_per_sqft if sqft > 0 else 5000
            
            return {
                'predicted_cost': round(estimated, 2),
                'confidence': 0.3,
                'method': 'avg_per_sqft',
                'training_samples': len(self.training_data)
            }
        
        # No data - rough estimate
        return {
            'predicted_cost': max(1000, sqft * 50),
            'confidence': 0.1,
            'method': 'no_data',
            'training_samples': 0
        }
    
    def get_statistics(self) -> Dict:
        """Get statistics from training data."""
        if not self.training_data:
            return {'error': 'No training data'}
        
        costs = [s['cost'] for s in self.training_data]
        widths = [s['width'] for s in self.training_data]
        heights = [s['height'] for s in self.training_data]
        sqfts = [s['sqft'] for s in self.training_data]
        
        # Count bay configurations
        bay_configs = {}
        for s in self.training_data:
            config = f"{s.get('bays_wide', 0)}x{s.get('bays_tall', 0)}"
            bay_configs[config] = bay_configs.get(config, 0) + 1
        
        # Sort by count
        common_configs = sorted(bay_configs.items(), key=lambda x: x[1], reverse=True)
        
        return {
            'sample_count': len(self.training_data),
            'avg_cost': round(sum(costs) / len(costs), 2),
            'min_cost': round(min(costs), 2),
            'max_cost': round(max(costs), 2),
            'avg_width': round(sum(widths) / len(widths), 0),
            'avg_height': round(sum(heights) / len(heights), 0),
            'avg_sqft': round(sum(sqfts) / len(sqfts), 1),
            'common_configurations': [
                {'config': c[0], 'count': c[1], 'percentage': round(c[1] / len(self.training_data) * 100, 1)}
                for c in common_configs[:5]
            ],
            'is_trained': self.is_trained
        }
    
    def get_status(self) -> Dict:
        """Get ML status."""
        # Reload data to ensure we have the latest count (without auto-training)
        self._load_data(auto_train=False)
        return {
            'training_samples': len(self.training_data),
            'is_trained': self.is_trained,
            'sklearn_available': ML_AVAILABLE
        }
    
    def remove_samples_by_project(self, project_name: str) -> int:
        """Remove all training samples for a given project.
        Returns the number of samples removed.
        """
        initial_count = len(self.training_data)
        self.training_data = [s for s in self.training_data if s.get('project') != project_name]
        removed_count = initial_count - len(self.training_data)
        
        if removed_count > 0:
            # Retrain if we still have enough data, otherwise clear model
            if len(self.training_data) >= 3 and ML_AVAILABLE:
                self._train()
            else:
                self.is_trained = False
                self.model = None
                self.scaler = None
            
            self._save_data()
            print(f"[ML] Removed {removed_count} training samples for project '{project_name}'")
        
        return removed_count
    
    def clear_data(self):
        """Clear all training data."""
        self.training_data = []
        self.is_trained = False
        self.model = None
        self.scaler = None
        self._save_data()


# Global instance
_predictor = None

def get_predictor() -> SimplePredictor:
    """Get or create the predictor instance."""
    global _predictor
    if _predictor is None:
        _predictor = SimplePredictor()
    return _predictor


# Public API
def add_project_to_training(elevation_data: Dict, project_name: str = "", elevation_name: str = "") -> bool:
    """Add an elevation to training data."""
    return get_predictor().add_sample(elevation_data, project_name, elevation_name)

def predict_project_cost(elevation_data: Dict) -> Dict:
    """Predict cost for an elevation."""
    return get_predictor().predict(elevation_data)

def train_ml_model() -> tuple:
    """Train the ML model.
    Returns (success: bool, message: str)
    """
    predictor = get_predictor()
    sample_count = len(predictor.training_data)
    
    if sample_count < 3:
        msg = f"Need minimum of 3 samples to train. Currently have {sample_count} sample(s)."
        print(f"[ML] {msg}")
        return False, msg
    
    success = predictor._train()
    if success:
        return True, f"Model trained successfully with {sample_count} samples!"
    else:
        return False, "Training failed. Please check the logs."

def get_training_status() -> Dict:
    """Get training status."""
    return get_predictor().get_status()

def get_pattern_insights() -> Dict:
    """Get pattern insights from training data."""
    return get_predictor().get_statistics()

def collect_training_data_from_projects() -> int:
    """Collect data from all existing projects."""
    predictor = get_predictor()
    collected = 0
    
    if not os.path.exists(PROJECTS_DIR):
        return 0
    
    for filename in os.listdir(PROJECTS_DIR):
        if filename.endswith('_Elevations.json'):
            project_name = filename.replace('_Elevations.json', '')
            filepath = os.path.join(PROJECTS_DIR, filename)
            
            try:
                with open(filepath, 'r') as f:
                    elevations = json.load(f)
                
                for elev_name, elev_data in elevations.items():
                    if predictor.add_sample(elev_data, project_name):
                        collected += 1
            except Exception as e:
                print(f"[ML] Error processing {filename}: {e}")
    
    print(f"[ML] Collected {collected} new samples")
    return collected

def remove_project_from_training(project_name: str) -> int:
    """Remove all training samples for a project.
    Returns the number of samples removed.
    """
    return get_predictor().remove_samples_by_project(project_name)

def remove_elevation_from_training(elevation_data: Dict, project_name: str = "", elevation_name: str = "") -> bool:
    """Remove a single elevation from training data.
    Returns True if removed, False if not found.
    """
    predictor = get_predictor()
    
    # Create same unique identifier as in add_sample
    width = elevation_data.get('opening_width_inches', 0)
    height = elevation_data.get('opening_height_inches', 0)
    bays_w = elevation_data.get('bays_wide', 0)
    bays_h = elevation_data.get('bays_tall', 0)
    finish = elevation_data.get('finish', '')
    
    sample_id = f"{project_name}_{elevation_name}_{width}_{height}_{bays_w}x{bays_h}_{finish}"
    
    initial_count = len(predictor.training_data)
    predictor.training_data = [s for s in predictor.training_data if s.get('id') != sample_id]
    
    if len(predictor.training_data) < initial_count:
        # Retrain if we still have enough data
        if len(predictor.training_data) >= 3 and ML_AVAILABLE:
            predictor._train()
        else:
            predictor.is_trained = False
            predictor.model = None
            predictor.scaler = None
        predictor._save_data()
        return True
    
    return False

def is_in_training(elevation_data: Dict, project_name: str = "", elevation_name: str = "") -> bool:
    """Check if an elevation is already in training data."""
    predictor = get_predictor()
    
    # Create same unique identifier as in add_sample
    width = elevation_data.get('opening_width_inches', 0)
    height = elevation_data.get('opening_height_inches', 0)
    bays_w = elevation_data.get('bays_wide', 0)
    bays_h = elevation_data.get('bays_tall', 0)
    finish = elevation_data.get('finish', '')
    
    sample_id = f"{project_name}_{elevation_name}_{width}_{height}_{bays_w}x{bays_h}_{finish}"
    
    for sample in predictor.training_data:
        if sample.get('id') == sample_id:
            return True
    
    return False

File: web/src/lib/ml-predictor.ts
/**
 * ML Cost Predictor - Browser-compatible linear regression
 * Matches Python's utils/ml_predictor.py behavior:
 *   4 features: [sqft, bays, width, height]
 *   3-tier prediction: exact match -> linear regression -> avg/sqft fallback
 *   Training data persisted to Supabase
 */

import { db } from './database';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface TrainingSample {
  id: string;
  project: string;
  elevation: string;
  width: number;
  height: number;
  bays: number;
  bays_wide: number;
  bays_tall: number;
  sqft: number;
  finish: string;
  cost: number;
  timestamp: string;
}

export interface PredictionResult {
  cost: number;
  confidence: number;
  method: 'exact_match' | 'ml_model' | 'avg_per_sqft' | 'no_data';
  training_samples: number;
}

export interface MLStatistics {
  sample_count: number;
  avg_cost: number;
  min_cost: number;
  max_cost: number;
  avg_width: number;
  avg_height: number;
  avg_sqft: number;
  avg_cost_per_sqft: number;
  common_configurations: Array<{
    config: string;
    count: number;
    percentage: number;
  }>;
}

export interface MLStatus {
  is_trained: boolean;
  sample_count: number;
  ml_available: boolean;
}

// ---------------------------------------------------------------------------
// Simple Linear Regression (no external deps)
// ---------------------------------------------------------------------------

class SimpleLinearRegression {
  private weights: number[] | null = null;
  private bias: number = 0;
  private featureMeans: number[] = [];
  private featureStds: number[] = [];
  private trained = false;

  /**
   * Train using ordinary least squares with feature standardization.
   * X: array of feature vectors [sqft, bays, width, height]
   * y: array of target values (cost)
   */
  train(X: number[][], y: number[]): void {
    if (X.length < 3) return;
    const n = X.length;
    const d = X[0].length;

    // Compute means and stds for standardization
    this.featureMeans = Array(d).fill(0);
    this.featureStds = Array(d).fill(0);

    for (let j = 0; j < d; j++) {
      let sum = 0;
      for (let i = 0; i < n; i++) sum += X[i][j];
      this.featureMeans[j] = sum / n;
    }

    for (let j = 0; j < d; j++) {
      let sum = 0;
      for (let i = 0; i < n; i++) {
        sum += (X[i][j] - this.featureMeans[j]) ** 2;
      }
      this.featureStds[j] = Math.sqrt(sum / n) || 1; // avoid div by zero
    }

    // Standardize features
    const Xs: number[][] = X.map(row =>
      row.map((val, j) => (val - this.featureMeans[j]) / this.featureStds[j]),
    );

    // OLS: w = (X^T X)^-1 X^T y (with bias via augmented matrix)
    // Augment with bias column
    const Xa = Xs.map(row => [...row, 1]);
    const dA = d + 1;

    // X^T X
    const XtX: number[][] = Array.from({ length: dA }, () => Array(dA).fill(0));
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < dA; j++) {
        for (let k = 0; k < dA; k++) {
          XtX[j][k] += Xa[i][j] * Xa[i][k];
        }
      }
    }

    // Add small ridge penalty for stability
    for (let j = 0; j < dA; j++) {
      XtX[j][j] += 1e-6;
    }

    // X^T y
    const Xty: number[] = Array(dA).fill(0);
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < dA; j++) {
        Xty[j] += Xa[i][j] * y[i];
      }
    }

    // Solve via Gauss-Jordan elimination
    const aug = XtX.map((row, i) => [...row, Xty[i]]);
    for (let j = 0; j < dA; j++) {
      // Pivot
      let maxRow = j;
      for (let i = j + 1; i < dA; i++) {
        if (Math.abs(aug[i][j]) > Math.abs(aug[maxRow][j])) maxRow = i;
      }
      [aug[j], aug[maxRow]] = [aug[maxRow], aug[j]];

      const pivot = aug[j][j];
      if (Math.abs(pivot) < 1e-12) continue;

      for (let k = j; k <= dA; k++) aug[j][k] /= pivot;
      for (let i = 0; i < dA; i++) {
        if (i === j) continue;
        const factor = aug[i][j];
        for (let k = j; k <= dA; k++) {
          aug[i][k] -= factor * aug[j][k];
        }
      }
    }

    const solution = aug.map(row => row[dA]);
    this.weights = solution.slice(0, d);
    this.bias = solution[d];
    this.trained = true;
  }

  predict(features: number[]): number | null {
    if (!this.trained || !this.weights) return null;
    // Standardize
    const xs = features.map(
      (val, j) => (val - this.featureMeans[j]) / this.featureStds[j],
    );
    let pred = this.bias;
    for (let j = 0; j < xs.length; j++) {
      pred += this.weights[j] * xs[j];
    }
    return Math.max(0, pred);
  }

  get isTrained(): boolean {
    return this.trained;
  }
}

// ---------------------------------------------------------------------------
// ML Predictor class
// ---------------------------------------------------------------------------

const MIN_SAMPLES = 3;

class MLPredictor {
  private model = new SimpleLinearRegression();
  private data: TrainingSample[] = [];
  private loaded = false;

  async loadData(): Promise<void> {
    if (this.loaded) return;
    try {
      const mlData = await db.getMLData();
      this.data = (mlData || []) as TrainingSample[];
      if (this.data.length >= MIN_SAMPLES) {
        this.trainModel();
      }
      this.loaded = true;
    } catch (e) {
      console.warn('Failed to load ML data:', e);
      this.data = [];
      this.loaded = true;
    }
  }

  private async saveData(): Promise<void> {
    try {
      await db.saveMLData(this.data);
    } catch (e) {
      console.warn('Failed to save ML data:', e);
    }
  }

  private trainModel(): void {
    if (this.data.length < MIN_SAMPLES) return;
    const X = this.data.map(s => [s.sqft, s.bays, s.width, s.height]);
    const y = this.data.map(s => s.cost);
    this.model = new SimpleLinearRegression();
    this.model.train(X, y);
  }

  async addSample(sample: Omit<TrainingSample, 'id' | 'timestamp'>): Promise<boolean> {
    const id = `${sample.project}_${sample.elevation}_${sample.width}_${sample.height}_${sample.bays_wide}x${sample.bays_tall}_${sample.finish}`;

    // Reject duplicates (matches Python behavior)
    if (this.data.some(s => s.id === id)) {
      return false;
    }

    const entry: TrainingSample = {
      ...sample,
      id,
      timestamp: new Date().toISOString(),
    };

    this.data.push(entry);

    if (this.data.length >= MIN_SAMPLES) {
      this.trainModel();
    }
    await this.saveData();
    return true;
  }

  async removeSample(id: string): Promise<void> {
    this.data = this.data.filter(s => s.id !== id);
    if (this.data.length >= MIN_SAMPLES) {
      this.trainModel();
    } else {
      this.model = new SimpleLinearRegression();
    }
    await this.saveData();
  }

  async removeSamplesByProject(project: string): Promise<void> {
    this.data = this.data.filter(s => s.project !== project);
    if (this.data.length >= MIN_SAMPLES) {
      this.trainModel();
    } else {
      this.model = new SimpleLinearRegression();
    }
    await this.saveData();
  }

  async clearAllSamples(): Promise<void> {
    this.data = [];
    this.model = new SimpleLinearRegression();
    await this.saveData();
  }

  /**
   * Check if an elevation is in training by building the full sample ID
   * (matches Python's is_in_training which checks exact ID match).
   */
  isInTraining(
    project: string,
    elevation: string,
    width?: number,
    height?: number,
    baysWide?: number,
    baysTall?: number,
    finish?: string,
  ): boolean {
    if (width != null && height != null && baysWide != null && baysTall != null && finish != null) {
      const id = `${project}_${elevation}_${width}_${height}_${baysWide}x${baysTall}_${finish}`;
      return this.data.some(s => s.id === id);
    }
    // Fallback: match by project + elevation name
    return this.data.some(s => s.project === project && s.elevation === elevation);
  }

  getSampleId(
    project: string,
    elevation: string,
    width?: number,
    height?: number,
    baysWide?: number,
    baysTall?: number,
    finish?: string,
  ): string | null {
    if (width != null && height != null && baysWide != null && baysTall != null && finish != null) {
      const id = `${project}_${elevation}_${width}_${height}_${baysWide}x${baysTall}_${finish}`;
      const found = this.data.find(s => s.id === id);
      return found?.id ?? null;
    }
    const found = this.data.find(s => s.project === project && s.elevation === elevation);
    return found?.id ?? null;
  }

  predict(
    width: number,
    height: number,
    baysWide: number,
    baysTall: number,
    sqft: number,
  ): PredictionResult {
    const bays = baysWide * baysTall || 1;

    // Tier 1: Exact match (within tolerance, exact bays like Python)
    for (const sample of this.data) {
      if (
        Math.abs(sample.width - width) < 1 &&
        Math.abs(sample.height - height) < 1 &&
        sample.bays === bays
      ) {
        return {
          cost: sample.cost,
          confidence: 0.95,
          method: 'exact_match',
          training_samples: this.data.length,
        };
      }
    }

    // Tier 2: ML model prediction
    if (this.model.isTrained) {
      const features = [sqft, bays, width, height];
      const predicted = this.model.predict(features);
      if (predicted != null && predicted > 0) {
        // Confidence formula matches Python: max(0.4, min(0.85, 1 - diff_pct * 0.5))
        const avgCost =
          this.data.reduce((s, d) => s + d.cost, 0) / this.data.length;
        const diffPct = avgCost > 0 ? Math.abs(predicted - avgCost) / avgCost : 1;
        const confidence = Math.max(0.4, Math.min(0.85, 1 - diffPct * 0.5));
        return {
          cost: Math.max(0, Math.round(predicted * 100) / 100),
          confidence: Math.round(confidence * 100) / 100,
          method: 'ml_model',
          training_samples: this.data.length,
        };
      }
    }

    // Tier 3: Average cost per sqft fallback (matches Python: max(1, sqft))
    if (this.data.length > 0) {
      const avgCostPerSqft =
        this.data.reduce((s, d) => s + d.cost / Math.max(1, d.sqft), 0) /
        this.data.length;
      const estimated = sqft > 0 ? sqft * avgCostPerSqft : 5000;
      return {
        cost: Math.round(estimated * 100) / 100,
        confidence: 0.3,
        method: 'avg_per_sqft',
        training_samples: this.data.length,
      };
    }

    // No data fallback
    return {
      cost: Math.max(1000, sqft * 50),
      confidence: 0.1,
      method: 'no_data',
      training_samples: 0,
    };
  }

  async train(): Promise<boolean> {
    if (this.data.length < MIN_SAMPLES) return false;
    this.trainModel();
    return this.model.isTrained;
  }

  getStatistics(): MLStatistics | null {
    if (this.data.length === 0) return null;
    const costs = this.data.map(d => d.cost);
    const avgCost = costs.reduce((s, v) => s + v, 0) / costs.length;
    const avgWidth = this.data.reduce((s, d) => s + d.width, 0) / this.data.length;
    const avgHeight = this.data.reduce((s, d) => s + d.height, 0) / this.data.length;
    const avgSqft = this.data.reduce((s, d) => s + d.sqft, 0) / this.data.length;
    const avgCostPerSqft = avgSqft > 0 ? avgCost / avgSqft : 0;

    // Common configurations
    const configCounts = new Map<string, number>();
    for (const sample of this.data) {
      const config = `${sample.bays_wide}x${sample.bays_tall} bays`;
      configCounts.set(config, (configCounts.get(config) ?? 0) + 1);
    }
    const commonConfigs = Array.from(configCounts.entries())
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5)
      .map(([config, count]) => ({
        config,
        count,
        percentage: (count / this.data.length) * 100,
      }));

    return {
      sample_count: this.data.length,
      avg_cost: avgCost,
      min_cost: Math.min(...costs),
      max_cost: Math.max(...costs),
      avg_width: avgWidth,
      avg_height: avgHeight,
      avg_sqft: avgSqft,
      avg_cost_per_sqft: avgCostPerSqft,
      common_configurations: commonConfigs,
    };
  }

  getStatus(): MLStatus {
    return {
      is_trained: this.model.isTrained,
      sample_count: this.data.length,
      ml_available: true,
    };
  }

  getTrainingData(): TrainingSample[] {
    return [...this.data];
  }
}

// ---------------------------------------------------------------------------
// Singleton
// ---------------------------------------------------------------------------

let _predictor: MLPredictor | null = null;

export function getPredictor(): MLPredictor {
  if (!_predictor) {
    _predictor = new MLPredictor();
  }
  return _predictor;
}
